controllers/machine_controller_node.go

```diff
@@ -0,0 +1,150 @@
+	logger := r.Log.WithValues("machine", machine.Name, "namespace", machine.Namespace)
+	// Check that the Machine hasn't been deleted or in the process.
+	if !machine.DeletionTimestamp.IsZero() {
+		return nil
+	}
+
+	logger = logger.WithValues("cluster", cluster.Name)
+
+	// Check that the Machine has a valid ProviderID.
+	if machine.Spec.ProviderID == nil || *machine.Spec.ProviderID == "" {
+		logger.Info("Machine doesn't have a valid ProviderID yet")
+		return nil
+	}
+
+	providerID, err := noderefutil.NewProviderID(*machine.Spec.ProviderID)
+	if err != nil {
+		return err
+	}
+
+	remoteClient, err := r.Tracker.GetClient(ctx, util.ObjectKey(cluster))
+	if err != nil {
+		return err
+	}
+
+	// Even if Status.NodeRef exists, continue to do the following checks to make sure Node is healthy
+	node, err := r.getNode(remoteClient, providerID)
+	if err != nil {
+		if err == ErrNodeNotFound {
+			// While a NodeRef is set in the status, getting ErrNodeNotFound error means the node is deleted.
+			// If Status.NodeRef is not set before, node still can be in the provisioning state.
+			if machine.Status.NodeRef != nil {
+				conditions.MarkFalse(machine, clusterv1.MachineNodeHealthyCondition, clusterv1.NodeNotFoundReason, clusterv1.ConditionSeverityError, "")
+			}
+			return errors.Wrapf(&capierrors.RequeueAfterError{RequeueAfter: 20 * time.Second},
+				"cannot assign NodeRef to Machine %q in namespace %q, no matching Node", machine.Name, machine.Namespace)
+		}
+		logger.Error(err, "Failed to assign NodeRef")
+		r.recorder.Event(machine, apicorev1.EventTypeWarning, "FailedSetNodeRef", err.Error())
+		return err
+	}
+
+	// Set the Machine NodeRef.
+	if machine.Status.NodeRef == nil {
+		machine.Status.NodeRef = &apicorev1.ObjectReference{
+			Kind:       node.Kind,
+			APIVersion: node.APIVersion,
+			Name:       node.Name,
+			UID:        node.UID,
+		}
+		logger.Info("Set Machine's NodeRef", "noderef", machine.Status.NodeRef.Name)
+		r.recorder.Event(machine, apicorev1.EventTypeNormal, "SuccessfulSetNodeRef", machine.Status.NodeRef.Name)
+		conditions.MarkFalse(machine, clusterv1.MachineNodeHealthyCondition, clusterv1.NodeProvisioningReason, clusterv1.ConditionSeverityWarning, "")
+		return nil
+	}
+
+	// Do the remaining node health checks, then set the node health to true if all checks pass.
+	if err := hasHealthyNodeConditions(node); err != nil {
+		conditions.MarkFalse(machine, clusterv1.MachineNodeHealthyCondition, clusterv1.NodeConditionsFailedReason, clusterv1.ConditionSeverityWarning, "Unhealthy conditions: %s", err.Error())
+		return nil
+	}
+
+	conditions.MarkTrue(machine, clusterv1.MachineNodeHealthyCondition)
+	return nil
+}
+
+// hasHealthyNodeConditions checks if a Node is healthy and does not have any semantically negative Conditions.
+// Returns error if NodeMemoryPressure, NodeDiskPressure, or NodePIDPressure Conditions are false or Ready Condition is true.
+func hasHealthyNodeConditions(node *apicorev1.Node) error {
```

Comment:
I personally found the signature of this func a little bit confusing, mostly because it is using an errors & kerrors.NewAggregate to compose a message.
What about 
```suggestion
func summarizeNodeConditions(node *apicorev1.Node) (status corev1.ConditionStatus, message string) {
```

Reply:
With this signature, we can only relay the first condition that fails. If we want to show all the failed conditions, error list makes sense in the sense that these are all errors on the Node side. Or I can use a string slice?

Reply:
The return message could be the concatenation of the failing conditions. 
I would avoid exposing a slice of messages

Reply:
I changed the signature like this but status has different meanings for different conditions, so to be consistent, returning false if a condition is false semantically.